src/main/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiff.scala

```diff
@@ -81,30 +81,30 @@ object AssertOpcDiff {
     }
   }
 
-  object AssertType extends Enumeration(1) {
-    val GteqZeroAssert, LteqAssert, LtInt64Assert, GtZeroAssert, EqAssert, IsCallerOriginAssert, IsSignerOriginAssert = Value
+  object AssertType extends Enumeration {
+    sealed case class AssertTypeVal(
+      assertType: Int,
+      operandCount: Int,
+      differ: (ExecutionContext, Array[Byte], Seq[DataEntry]) => Either[ValidationError, OpcDiff])
+    extends Val(assertType) { def *(n: Int): Int = n * assertType }
+
+    val GteqZeroAssert       = AssertTypeVal(1, 1, (c, b, d) => gtEq0(d(b(1))))
+    val LteqAssert           = AssertTypeVal(2, 2, (c, b, d) => ltEq(d(b(1)), d(b(2))))
+    val LtInt64Assert        = AssertTypeVal(3, 1, (c, b, d) => ltInt64(d(b(1))))
+    val GtZeroAssert         = AssertTypeVal(4, 1, (c, b, d) => gt0(d(b(1))))
+    val EqAssert             = AssertTypeVal(5, 2, (c, b, d) => equql(d(b(1)), d(b(2))))
+    val IsCallerOriginAssert = AssertTypeVal(6, 1, (c, b, d) => isCallerOrigin(c)(d(b(1))))
+    val IsSignerOriginAssert = AssertTypeVal(7, 1, (c, b, d) => isSignerOrigin(c)(d(b(1))))
+
+    def fromByte(b: Byte): Option[AssertTypeVal] = Try(AssertType(b).asInstanceOf[AssertTypeVal]).toOption
   }
 
-  def parseBytes(context: ExecutionContext)
-                (bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, OpcDiff] = {
-    if (checkAssertDataIndex(bytes, data.length)) {
-      (bytes.headOption.flatMap(f => Try(AssertType(f)).toOption), bytes.length) match {
-        case (Some(AssertType.GteqZeroAssert), 2) => gtEq0(data(bytes(1)))
-        case (Some(AssertType.LteqAssert), 3) => ltEq(data(bytes(1)), data(bytes(2)))
-        case (Some(AssertType.LtInt64Assert), 2) => ltInt64(data(bytes(1)))
-        case (Some(AssertType.GtZeroAssert), 2) => gt0(data(bytes(1)))
-        case (Some(AssertType.EqAssert), 3) => eq(data(bytes(1)), data(bytes(2)))
-        case (Some(AssertType.IsCallerOriginAssert), 2) => isCallerOrigin(context)(data(bytes(1)))
-        case (Some(AssertType.IsSignerOriginAssert), 2) => isSignerOrigin(context)(data(bytes(1)))
-        case _ => Left(ContractInvalidOPCData)
-      }
+  // index out of bound exception if custom function with this opc in `data(byte(i))`
+  override def parseBytesDf(context: ExecutionContext)(bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, OpcDiff] =
+    bytes.headOption.flatMap(AssertType.fromByte(_)) match {
+      case Some(t: AssertType.AssertTypeVal) if bytes.length == t.operandCount + 1 =>
+        t.differ(context, bytes, data)
+      case _ => Left(ContractInvalidOPCData)
```

Comment:
is there any place checked the value of byte in `bytes` bounded by the `data` length. since the value in `bytes` represent an index of `data`.

Reply:
oh, i forgot to use `checkData` in this object